x/stake/keeper/val_state_change.go

```diff
@@ -0,0 +1,262 @@
+package keeper
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/stake/types"
+)
+
+// Apply and return accumulated updates to the bonded validator set
+//
+// CONTRACT: Only validators with non-zero power or zero-power that were bonded
+// at the previous block height or were removed from the validator set entirely
+// are returned to Tendermint.
+func (k Keeper) ApplyAndReturnValidatorSetUpdates(ctx sdk.Context) (updates []abci.ValidatorUpdate) {
+
+	store := ctx.KVStore(k.storeKey)
+	maxValidators := k.GetParams(ctx).MaxValidators
+
+	// retrieve last validator set
+	last := k.retrieveLastValidatorSet(ctx)
+
+	// iterate over validators, highest power to lowest
+	iterator := sdk.KVStoreReversePrefixIterator(store, ValidatorsByPowerIndexKey)
+	count := 0
+	for ; iterator.Valid() && count < int(maxValidators); iterator.Next() {
+
+		// fetch the validator
+		operator := sdk.ValAddress(iterator.Value())
+		validator := k.mustGetValidator(ctx, operator)
+
+		if validator.Jailed {
+			panic("should never retrieve a jailed validator from the power store")
+		}
+
+		// if we get to a zero-power validator (which we don't bond),
+		// there are no more possible bonded validators
+		// note: we must check the ABCI power, since we round before sending to Tendermint
+		if validator.Tokens.RoundInt64() == int64(0) {
+			break
+		}
+
+		// apply the appropriate state change if necessary
+		switch validator.Status {
+		case sdk.Unbonded:
+			validator = k.unbondedToBonded(ctx, validator)
+		case sdk.Unbonding:
+			validator = k.unbondingToBonded(ctx, validator)
+		case sdk.Bonded:
+			// no state change
+		}
+
+		// fetch the old power bytes
+		var opbytes [sdk.AddrLen]byte
+		copy(opbytes[:], operator[:])
+		oldPowerBytes, ok := last[opbytes]
+
+		// calculate the new power bytes
+		newPowerBytes := validator.ABCIValidatorPowerBytes(k.cdc)
+
+		// update the validator set if power has changed
+		if !ok || !bytes.Equal(oldPowerBytes, newPowerBytes) {
+			updates = append(updates, validator.ABCIValidatorUpdate())
+		}
+
+		// validator still in the validator set, so delete from the copy
+		delete(last, opbytes)
+
+		// set the bonded validator index
+		store.Set(GetBondedValidatorIndexKey(operator), newPowerBytes)
+
+		// keep count
+		count++
+
+	}
+
+	// sort the no-longer-bonded validators
+	noLongerBonded := k.sortNoLongerBonded(last)
+
+	// iterate through the sorted no-longer-bonded validators
+	for _, operator := range noLongerBonded {
+
+		// fetch the validator
+		validator := k.mustGetValidator(ctx, sdk.ValAddress(operator))
+
+		// bonded to unbonding
+		k.bondedToUnbonding(ctx, validator)
+
+		// remove validator if it has no more tokens
+		if validator.Tokens.IsZero() {
+			k.RemoveValidator(ctx, validator.OperatorAddr)
+		}
+
+		// delete from the bonded validator index
+		store.Delete(GetBondedValidatorIndexKey(operator))
```

Comment:
see my comment on the `RemoveValidator` func - maybe we should move this in there?

Reply:
https://github.com/cosmos/cosmos-sdk/pull/2394#discussion_r222291735